Feature/wp lib

iruzevic opened the pull request:
- Removing a lot of stuff that is not part of boilerplate like utilities and etc. this should be in wiki
- Changing folder structure to src
- Implementing Eightshift-libs repo currently from github
- Updating docs
- Adding type tinting.
- Updating npm packages

IvanGrginovInf commented on composer.json:
```diff
@@ -21,7 +24,7 @@
-    "check-cs": "@php ./vendor/bin/phpcs --extensions=php --ignore=/vendor/",
-    "fix-cs": "@php ./vendor/bin/phpcbf --extensions=php --ignore=/vendor/"
+    "check-cs": "@php ./vendor/bin/phpcs",
+    "fix-cs": "@php ./vendor/bin/phpcbf"
```
Why we no longer `--ignore=/vendor/` by default?

iruzevic replied:
Denis told me that this is implemented in the sniffer

dingo-d replied:
No since this is defined in the standards 😉

IvanGrginovInf replied:
@dingo-d you mean you can tell phpcs to use specific settings for sniffing when using a specific standard?

iruzevic replied:
yes that is what phpcs.xml.dist is used

dingo-d replied:
The `node_modules` `vendor` and some are disabled by default in the Infinum's standards, but you can also add (and should for every project) project specific `phpcs.xml.dist` file where you define the standard you use (in our case our own) plus some project specific overrides 🙂

dingo-d commented on composer.json:
```diff
@@ -3,15 +3,18 @@
     "dealerdirect/phpcodesniffer-composer-installer": "^0.5"
   },
+  "require": {
+    "infinum/eightshift-libs": "dev-develop"
```
This will have to be changed once the lib is out...

iruzevic replied:
yes once we deploy lib as public I will change this

dingo-d commented on readme.txt:
```diff
@@ -1,9 +1,9 @@
-Requires PHP: 7
-Tested up to: 4.9.8
+Requires PHP: 7.0
+Tested up to: 5.0.3
```
5.1.1?

iruzevic replied:
👍

dingo-d commented on readme.txt:
```diff
@@ -18,7 +18,7 @@ This is an open-source modern boilerplate / starter theme with all the latest de
 
 It is intended for users with at least some dev experience (php and js) and it's a great starting point if you want to build a theme for your own or client's project, to release a theme on WordPress.org or to sell it.
```
Change the `It` to `Eightshift WordPress Boilerplate` for more clarity 🙂

iruzevic replied:
👍

dingo-d commented on src/admin/class-admin.php:
```diff
@@ -0,0 +1,97 @@
+   * @since 1.0.0
+   */
+  public function register() : void {
+    \add_action( 'login_enqueue_scripts', [ $this, 'enqueue_styles' ] );
```
Just wondering, should we force load from global namespace. Could become difficult to mock if we decide to later on (just thinking out loud). If it doesn't throw an error when we run it (just `add_action` instead of `\add_action`), we can use the unslashed version.

iruzevic replied:
👍

dingo-d commented on src/admin/class-media.php:
```diff
@@ -8,36 +8,59 @@
+ * This class handles all media options. Sizes, Types, Features, etc.
  */
-class Media {
+class Media implements Service {
```
I'd like to leave this, but if we'd want to publish a theme to wp.org, I'm 99% sure the theme wouldn't pass, because of the svg upload enabling (sounds like a plugin territory). So we'd need to document this properly for anyone who wants to use it for this purpose.

iruzevic replied:
let's agree this is going to be stupid boilerplate without forcing anything to the user.

All this featurs will be in our BP and WIKI

dingo-d commented on src/admin/menu/class-bem-menu-walker.php:
```diff
@@ -76,7 +78,7 @@ public function display_element( $element, &$children_elements, $max_depth, $dep
    * @since 1.0.0
    */
-  public function start_lvl( &$output, $depth = 1, $args = array() ) { // phpcs:ignore Generic.CodeAnalysis.UnusedFunctionParameter.FoundInExtendedClassAfterLastUsed
+  public function start_lvl( &$output, $depth = 1, $args = [] ) { // phpcs:ignore Generic.CodeAnalysis.UnusedFunctionParameter.FoundInExtendedClassAfterLastUsed
```
The comment should be added to the `phpcs.xml.dist`

```xml
<rule ref="Generic.CodeAnalysis.UnusedFunctionParameter">
  <severity>0</severity>
</rule>
```
Should silence all these errors.

iruzevic replied:
👍 tnx

dingo-d commented on src/admin/menu/class-bem-menu-walker.php:
```diff
@@ -108,7 +110,7 @@ public function start_lvl( &$output, $depth = 1, $args = array() ) { // phpcs:ig
    * @since 1.0.0
    */
-  public function start_el( &$output, $item, $depth = 0, $args = array(), $id = 0 ) {  // phpcs:ignore Generic.CodeAnalysis.UnusedFunctionParameter.FoundInExtendedClassAfterLastUsed
+  public function start_el( &$output, $item, $depth = 0, $args = [], $id = 0 ) {  // phpcs:ignore Generic.CodeAnalysis.UnusedFunctionParameter.FoundInExtendedClassAfterLastUsed
```
⬆️

iruzevic replied:
👍

dingo-d commented on src/admin/menu/class-bem-menu-walker.php:
```diff
@@ -120,7 +122,7 @@ public function start_el( &$output, $item, $depth = 0, $args = array(), $id = 0
 
     if ( ! empty( $item->classes ) ) {
-      $user_classes = array();
+      $user_classes = [];
```
The below `foreach` can be refactored with `array_map` 🙂

iruzevic replied:
this will be refactored by Tihomir so I will leave it for him

dingo-d commented on src/class-manifest.php:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Class Mainfest
+ */
+class Manifest extends LibManifest implements Service {
```
You need to implement `Service` here? I thought this was handled in the lib? 🤔 Can you test and see if something fails? 🙂

iruzevic replied:
Lib only checks if the manifest.json exists and processes the file and returns it as a method.

In your project (boilerplate) you should assign that processed data to the global variable and use it as you like.

I think this gives us the best flexibility for the future

dingo-d commented on src/helpers/class-general-helper.php:
```diff
@@ -23,6 +23,8 @@ class General_Helper {
+   *
+   * // TODO: Refactor and remove.
    */
   public static function get_array_value( $key, $array ) {
```
I thought that we said this wasn't needed? We'd use null coalesce operator instead (`??`) or were there some cases where we couldn't use this.

CC @IvanGrginovInf

IvanGrginovInf replied:
@dingo-d Yup, hence the `to do` I assume. (refactor all places where this is used and remove).

iruzevic replied:
removed

dingo-d commented on src/theme/class-theme.php:
```diff
@@ -0,0 +1,95 @@
+     * This will remove some default functionality, but it mostly removes unnecessary
+     * meta tags from the head section.
+     */
+    \remove_action( 'wp_head', 'print_emoji_detection_script', 7 );
```
I'll leave this, but same as with svg, leave a notice that such a theme wouldn't pass wp.org review because there is a rule that forbids removing core actions.

iruzevic replied:
👍

iruzevic commented:
I have:

- Removed general helper
- Removed object helper
- Removed svg media hooks
- Removed all remove_action hooks
- Changed manifest to general folder
- Added new class for shortcodes.
- Created Changelogs

iruzevic commented:
I have updated the webpack build again to optimize for the multiple configs
